[move] Adds Math::sqrt()

damirka opened the pull request.

sblackshear commented on sui_programmability/framework/sources/Math.move:
```diff
@@ -21,4 +21,22 @@ module Sui::Math {
         }
     }
+
+    /// Get Square Root for `x`
```
I think tit would be useful for this to explain what happens when the result is fractional.

kchalkias reviewed (dismissed):
We should add some extra unit tests for large numbers, just to ensure correctness across the range of `u64`.

huitseeker reviewed (commented):
Have you looked at `proptest`?

https://altsysrq.github.io/proptest-book/intro.html

It would let you fuzz your impl (with shrinking!) with something as simple as:

```
proptest! {
  fn test_sqrt(
            num in any::<u64>()
        ) {
            ... // your test here
    }
}
```

sblackshear commented:
> Have you looked at `proptest`?
> 
> https://altsysrq.github.io/proptest-book/intro.html
> 
> It would let you fuzz your impl (with shrinking!) with something as simple as:
> 
> ```
> proptest! {
>   fn test_sqrt(
>             num in any::<u64>()
>         ) {
>             ... // your test here
>     }
> }
> ```

Fuzzing support in Move unit testing would be great for us to look into at some point (and good for parity w/ Foundry)

damirka commented:
> We should add some extra unit tests for large numbers, just to ensure correctness across the range of `u64`.

@kchalkias I've added support for large numbers (also added a test for U64::MAX).

kchalkias reviewed (approved):
LGTM, a reference to the actual algorithm (where did we find this logic?) would help for extra validation in the future. Due to the absence of `prop_tests` in Move we could also add to our codebase a prop_test for its Rust version:
```
fn sqrt_u64(num: u64) -> u64 {
    let mut bit = 1u128 << 64;
    let mut res = 0u128;
    let mut x = num as u128;

    while bit != 0 {
        if x >= res + bit {
            x = x - (res + bit);
            res = (res >> 1) + bit;
        } else {
            res = res >> 1;
        };
        bit = bit >> 2;
    };

    res as u64
}
```

huitseeker commented:
> LGTM, a reference to the actual algorithm (where did we find this logic?) would help for extra validation in the future.

The actual original reference is:

Woo, C.C. _The Fundamental Operations in Bead Arithmetic. How to Use the Chinese Abacus_, China Lace Co., Ltd., Hong Kong, ca. 1930, Chapter 7:
https://www.jumpjet.info/Offbeat-Internet/More/Misc/The_Fundamental_Operations_in_Bead_Arithmetic.pdf

huitseeker commented:
Or it may be August Toepler, 1865, in
Reuleaux, F.: _Prof. Toepler’s Verfahren der Wurzelausziehung mittelst der Thomas’schen Rechenmaschine._ In: Polytechnisches Journal, 179. Bd., Jg. 1866, S. 260–264
http://rechnerlexikon.de/files/PolytJournal1866-Toepler.pdf

Explanation: https://www.researchgate.net/publication/356875243_Toepler's_Algorithm_to_Calculate_Square_Roots_with_Mechanical_Calculators_Mathematical_background_and_practical_demonstration

huitseeker commented:
Tracing the source of this will take you from:
http://medialab.freaknet.org/martin/src/sqrt/
to :
https://cs.stackexchange.com/questions/52683/where-can-i-find-an-original-reference-for-this-integer-square-root-algorithm
which points at this 1930 reference:
https://www.jumpjet.info/Offbeat-Internet/More/Misc/The_Fundamental_Operations_in_Bead_Arithmetic.pdf
then through :
https://computarium.lcd.lu/literature/HISTORIC_CALCULATING/Square_Root/Toepler/square_root_revisited.htm
it will lead you to :
https://www.researchgate.net/publication/356875243_Toepler's_Algorithm_to_Calculate_Square_Roots_with_Mechanical_Calculators_Mathematical_background_and_practical_demonstration
which hints at the original source being this 1866 paper:
http://rechnerlexikon.de/files/PolytJournal1866-Toepler.pdf

Conclusion: this is at least over 150 y.o.

kchalkias commented:
> it's probably battle tested then
